Loop/Managers/Alerts/AlertStore.swift

```diff
@@ -0,0 +1,150 @@
+//
+//  AlertStore.swift
+//  Loop
+//
+//  Created by Rick Pasetto on 5/11/20.
+//  Copyright © 2020 LoopKit Authors. All rights reserved.
+//
+
+import CoreData
+import LoopKit
+
+public class AlertStore {
+    
+    public enum AlertStoreError: Error {
+        case notFound
+    }
+    
+    // Available for tests only
+    let managedObjectContext: NSManagedObjectContext
+
+    private let persistentContainer: NSPersistentContainer
+        
+    private let log = DiagnosticLog(category: "AlertStore")
+
+    private let dataAccessQueue = DispatchQueue(label: "com.loop.AlertStore.dataAccessQueue", qos: .utility)
+    
+    public init(storageFileURL: URL? = nil) {
+        managedObjectContext = NSManagedObjectContext(concurrencyType: .privateQueueConcurrencyType)
+        managedObjectContext.mergePolicy = NSMergeByPropertyStoreTrumpMergePolicy
+        managedObjectContext.automaticallyMergesChangesFromParent = true
+
+        let storeDescription = NSPersistentStoreDescription()
+        if let storageFileURL = storageFileURL {
+            storeDescription.url = storageFileURL
+        } else {
+            storeDescription.type = NSInMemoryStoreType
+        }
+        persistentContainer = NSPersistentContainer(name: "AlertStore")
+        persistentContainer.persistentStoreDescriptions = [storeDescription]
+        persistentContainer.loadPersistentStores { _, error in
+            if let error = error {
+                fatalError("Unable to load persistent stores: \(error)")
+            }
+        }
+        managedObjectContext.persistentStoreCoordinator = persistentContainer.persistentStoreCoordinator
+    }
+}
+
+// MARK: Alert Recording
+
+extension AlertStore {
+    
+    public func recordIssued(alert: DeviceAlert, at date: Date = Date(), completion: ((Result<Void, Error>) -> Void)? = nil) {
+        dataAccessQueue.async {
+            self.managedObjectContext.perform {
+                _ = StoredAlert(from: alert, context: self.managedObjectContext, issuedDate: date)
+                do {
+                    try self.managedObjectContext.save()
+                    self.log.default("Recorded alert: %{public}@", alert.identifier.value)
+                    completion?(.success)
+                } catch {
+                    self.log.error("Could not store alert: %{public}@", String(describing: error))
+                    completion?(.failure(error))
+                }
+            }
+        }
+    }
+    
+    public func recordAcknowledgement(of identifier: DeviceAlert.Identifier, at date: Date = Date(),
+                                      completion: ((Result<Void, Error>) -> Void)? = nil) {
```

Comment:
The `(Result<Void, Error>)` is interesting. We have typically just used `Error?` in this case.  They are very close, type-wise. Seems like there are tradeoffs both ways. It's somewhat more explicit to use the Result type, but you lose a little syntactic shortcut like `if let error = error` in handling the response (It becomes`if case .error(let error) = error`).

Reply:
I prefer `Result` because:
1) it forces callers to think about the error case.  It is much easier to ignore the `Error?` passed in, but the wrapper `Result` forces you to unwrap it to get the "actual" result.
2) I guess IMHO `if let error = error` and `if case .error(let error) = error` are both weird.  That's why (you'll notice) I prefer a `switch` statement.  
3) Using the `Void` version (`Result<Void, Error>`) gets you used to the pattern, so the code looks the same as the non-`Void` versions.  (It's handy to have the 
```swift
extension Result where Success == Void {
    static var success: Result {
        return Result.success(Void())
    }
}
```
code below, too.)

WDYT?

Reply:
Yeah, I think it's fine. It seems a little less idiomatic, but I don't feel strongly about it. Just saw it was a new pattern and thought it was interesting.
I think you can ignore an Optional (which is actually very similar to a Result) just as easily, and can unwrap it the same, but Optional has a little more built in syntax support.

Reply:
I had been using my own `Result` type for a while before Swift officially added it. I took it as a cue that this was a better way to go, and emboldened me to eschew the `Error?` route.

I'm not sure I follow your counterargument about ignoring an Optional... that's what my (1) argument above...?

As far as syntax support... I'm not sure if there are plans for nicer syntax support in Swift.

Reply:
I was just saying I think it's just as easy to ignore an `Error?` type result as it is a `Result<Void, Error>` type, so I don't think argument #1 applies. I love using `Result` when there is an actual success type (not Void) that callers may need.  And as far as syntax niceties, I'm just talking about `Optionals`, even though they're a simple enum like `Result`, having some syntatic shortcuts, like `if let`, optional chaining, unwrapping, etc. 

I'm not opposed to `Result<Void, Error>` at all, and may use it if I feel the explicitness helps a particular case. It seems a little more heavy than `Error?`, but sometimes being more explicit is nice.